src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -21880,6 +21880,548 @@ void M(string? s)
+class D
+{
+    static void M(C c, string? p1, string? p2)
+    {
+        _ = p1 ?? throw null!;
+        string? x1 = null;
+        c.M(p1, ref x1);
+        x1.ToString();
+
+        string? x2 = null;
+        c.M(p2, ref x2);
+        x2.ToString(); // 1
+
+        string? x3 = """";
+        c.M(p2, ref x3);
+        x3.ToString(); // 2
+    }
+}";
+            var expected = new[]
+            {
+                // (13,9): warning CS8602: Dereference of a possibly null reference.
+                //         x2.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x2").WithLocation(13, 9),
+                // (17,9): warning CS8602: Dereference of a possibly null reference.
+                //         x3.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x3").WithLocation(17, 9)
+            };
+
+            var lib = CreateNullableCompilation(new[] { NotNullIfNotNulltAttributeDefinition, lib_cs });
+            var comp = CreateNullableCompilation(source, references: new[] { lib.EmitToImageReference() });
+            comp.VerifyDiagnostics(expected);
+
+            var comp2 = CreateNullableCompilation(new[] { source, lib_cs, NotNullIfNotNulltAttributeDefinition });
+            comp2.VerifyDiagnostics(expected);
+            CompileAndVerify(comp2);
+        }
+
+        [Fact]
+        public void NotNullIfNotNull_ByValueParameter()
+        {
+            var lib_cs =
+@"using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    public void M(string? p, [NotNullIfNotNull(""p"")] string? p2) => throw null!;
+}";
+            var source = @"
+class D
+{
+    static void M(C c, string? p1, string? p2)
+    {
+        _ = p1 ?? throw null!;
+        string? x1 = null;
+        c.M(p1, x1);
+        x1.ToString(); // 1
+
+        string? x2 = null;
+        c.M(p2, x2);
+        x2.ToString(); // 2
+
+        string? x3 = """";
+        c.M(p2, x3);
+        x3.ToString(); // 3
+    }
+}";
+            var expected = new[]
+            {
+                // (9,9): warning CS8602: Dereference of a possibly null reference.
+                //         x1.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x1").WithLocation(9, 9),
+                // (13,9): warning CS8602: Dereference of a possibly null reference.
+                //         x2.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x2").WithLocation(13, 9)
+            };
+
+            var lib = CreateNullableCompilation(new[] { NotNullIfNotNulltAttributeDefinition, lib_cs });
+            var comp = CreateNullableCompilation(source, references: new[] { lib.EmitToImageReference() });
+            comp.VerifyDiagnostics(expected);
+
+            var comp2 = CreateNullableCompilation(new[] { source, lib_cs, NotNullIfNotNulltAttributeDefinition });
+            comp2.VerifyDiagnostics(expected);
+            CompileAndVerify(comp2);
+        }
```

Comment:
>} [](start = 8, length = 1)

Are we testing type parameters? For instance:
```
[return: NotNullIfNotNull("u")] static T F<T, U>(U u) { ... }
```

Reply:
I'll add some tests, but since the mechanism is the same as used for [NotNull], I don't plan to do many permutations.